Implement Quiz Reducer, Update Quiz Editor, Create Quiz Details & Quiz Preview Pages

allisonnblairr opened the pull request:
added quiz reducer to keep track of quiz state, updated quiz editor and quiz list to use the reducer, added restrictions for faculty and student views, added quiz details page with all requirements, added dummy quiz preview page that uses some mock data to traverse through questions

tigerhippo commented on src/Kambaz/Courses/Quizzes/QuizDetails.tsx:
```diff
@@ -1,3 +1,89 @@
+              <div><strong>Assignment Group:</strong> {quiz.assignmentGroup || 'QUIZZES'} </div>
+              <div><strong>Shuffle Answers:</strong> {quiz.shuffleAnswers === 'true' ? 'Yes' : 'No'}</div>
+              <div><strong>Time Limit:</strong> {quiz.timeLimit || '20'} Minutes</div>
+              <div><strong>Multiple Attempts:</strong> {quiz.hasMultipleAttempts === 'true' ? 'Yes' : 'No'}</div>
```
In addition to whether this quiz allows multiple attempts, we also need to show how many attempts this quiz has (make this another line called 'Number of Attempts')

tigerhippo commented on src/Kambaz/Courses/Quizzes/QuizDetails.tsx:
For some reason, the fields from my saved quiz are not being reflected in the details. For example, I made a quiz with the Quiz Type: Practice Quiz, but on the details page, it shows Quiz Type: Graded Quiz. I think this might be a reducer issue.

tigerhippo commented on src/Kambaz/Courses/Quizzes/QuizDetails.tsx:
When I click 'Edit' from the details page, the fields from my saved quiz are not reflected in the quiz editor either.

tigerhippo commented on src/Kambaz/Courses/Quizzes/index.tsx:
When I click on an individual quiz in the quiz list, the instructions for that quiz are not showing up in the quiz editor (the instructions that I wrote for that quiz probably weren't saved).

tigerhippo commented on src/Kambaz/Courses/Quizzes/index.tsx:
I currently can't change the unpublished status to a published status when clicking on the publish/unpublish symbol in the quiz list page.

tigerhippo commented on src/Kambaz/Courses/Quizzes/QuizEditor.tsx:
When I click the publish button and then 'save', this isn't being reflected in the quiz list or when I open up the same quiz again.

tigerhippo commented on src/Kambaz/Courses/Quizzes/index.tsx:
I currently can't change the unpublished status to a published status when clicking on the publish/unpublish menu option (when you click on the three dots next to each quiz).

tigerhippo commented on src/Kambaz/Courses/Quizzes/QuizDetails.tsx:
When I click on a quiz in the quiz list page, and then edit some fields, and then click 'save', the details page doesn't show the correct quiz type. The other fields on the details page seems to be correct though, and the fields on the preview page seem to be correct also.

tigerhippo reviewed (commented):
Overall it looks good! The main things I commented on were the fact that when I save a new quiz, the field values that I wrote for that quiz are not being reflected in the other pages. For some reason, when I click on an existing quiz and edit the fields values, most of the values are being reflected in the other pages. Try to manually test different cases to see what's happening with the fields (there are many cases lmao). I also reviewed your PR based on the demo checklist (I wrote out all the things you did successfully in green and things that aren't completed yet in red - I wrote this in the google doc under your tasks).

allisonnblairr commented:
I have checked all pr comments and implemented all fixes. I have also added extra stuff like showing quiz preview for students (minus editing buttons) I have also implemented all proper disabling and enabling of elements based on if you are faculty or a sudent

tigerhippo commented:
Looks good! One last thing is the current default values for some fields in the quiz editor are incorrect, they should be changed to:
1) Shuffle Answers: Yes
2) Time Limit: Yes, Minutes: 20
3) One Question at a Time: Yes
4) Number of Attempts: 1

tigerhippo commented:
The default values for some fields in the quiz editor are incorrect (after clicking +Quiz), they should be changed to:

1) Shuffle Answers: Yes
2) Time Limit: Yes, Minutes: 20
3) One Question at a Time: Yes
4) Number of Attempts: 1

These fields seem to be correct after you click 'Edit' in the quiz details page, which is interesting.

tigerhippo commented:
When I edit an existing quiz to have the following field values:
Time Limit: No, Minutes: blank 

Then I click edit again, the field values are now:
Time Limit: No, Minutes: 20

allisonnblairr commented:
> The default values for some fields in the quiz editor are incorrect (after clicking +Quiz), they should be changed to:
> 
> 1. Shuffle Answers: Yes
> 2. Time Limit: Yes, Minutes: 20
> 3. One Question at a Time: Yes
> 4. Number of Attempts: 1
> 
> These fields seem to be correct after you click 'Edit' in the quiz details page, which is interesting.

This is known - this will be fixed once back-end is implemented since default values are handled by the schema.
